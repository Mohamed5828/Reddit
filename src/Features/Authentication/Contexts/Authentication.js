import { createContext, useContext } from "react";

import useLocalStorage from "./useLocalStorage";

/* Creating a context object. */
const AuthContext = createContext(null);

/**
 * The AuthProvider is a function that takes in a child component and returns a context provider that
 * has a value of an object with the following properties: user, login, logout, getToken,
 * getExpirationDate, isLoggedIn, getFullName.
 * @returns The AuthContext.Provider is being returned.
 */
export const AuthProvider = ({ children }) => {
  // Local storage for user information
  const [user, setUser] = useLocalStorage("user", "null");

  /**
   * The login function takes a user object as an argument, and then sets the user state to the
   * stringified version of the user object.
   * @param user - The user object that is returned from the login function.
   */
  const login = async (user) => {
    setUser(JSON.stringify(user));
  };

  // Handle logout
  const logout = () => {
    localStorage.clear();
    setUser(null);
  };

  // Return user's full name
//   const getFullName = () => {
//     return JSON.parse(user).fullName;
//   };

  // Return user's token
  const getToken = () => {
    return JSON.parse(user).accessToken;
  };

  // Return token's expiration date
  const getExpirationDate = () => {
    return JSON.parse(user).expiresIn;
  };

  // Is user logged in?
  const isLoggedIn = () => {
    return user !== "null";
  };
  return (
    <AuthContext.Provider
      value={{
        //user,
        login,
        logout,
        getToken,
        getExpirationDate,
        isLoggedIn,
        //getFullName,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};

/**
 * It returns the value of the AuthContext object.
 * @returns The AuthContext object.
 */
export const useAuth = () => {
  return useContext(AuthContext);
};
