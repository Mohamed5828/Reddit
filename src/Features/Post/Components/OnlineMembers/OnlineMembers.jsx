// Import components
import OverlayTrigger from "react-bootstrap/OverlayTrigger";
import Tooltip from "react-bootstrap/Tooltip";

// Import styled components
import {
  Container,
  Members,
  Online,
  OnlineMembersContainer,
} from "./OnlineMembers.styled";
/**
 * component that show number of members
 * and number of online members
 *
 * @returns {React.Component} Online Members component
 */
const OnlineMembers = () => {
  return (
    <Container>
      <OnlineMembersContainer>
        <OverlayTrigger
          placement={"bottom"}
          delay="200"
          overlay={<Tooltip>{`5 Members`}</Tooltip>}
        >
          <Members>
            <div>{5}</div>
            <p>Members</p>
          </Members>
        </OverlayTrigger>
      </OnlineMembersContainer>
      <OnlineMembersContainer>
        <OverlayTrigger
          placement={"bottom"}
          overlay={<Tooltip>{"6 Online"}</Tooltip>}
          delay="200"
        >
          <Online>
            <div>{6}</div>
            <p>Online</p>
          </Online>
        </OverlayTrigger>
      </OnlineMembersContainer>
    </Container>
  );
};

export default OnlineMembers;
