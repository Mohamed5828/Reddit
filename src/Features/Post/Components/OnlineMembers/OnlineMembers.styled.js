import styled from "styled-components";

export const Container = styled.div`
  display: grid;
  grid-template-columns: auto auto;
  font-size: 12px;
`;

export const Members = styled.div`
  height: fit-content;

  * {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
  }
  div {
    font-size: 16px;
    font-weight: 500;
  }

  p {
    font-weight: 400;
    color: ${({ theme }) => theme.color.muted};
    font-size: 12px;
    display: inline-block;
    word-break: break-word;
    margin: 0;
  }
`;

export const Online = styled.div`
  height: fit-content;

  div {
    font-size: 16px;
    font-weight: 500;
    width: fit-content;
  }

  div::before {
    content: "●";
    margin-right: 4px;
    color: #46d160;
  }

  p {
    font-weight: 400;
    color: ${({ theme }) => theme.color.muted};
    font-size: 12px;
    display: inline-block;
    word-break: break-word;
    margin: 0;
    width: fit-content;
  }
`;

export const OnlineMembersContainer = styled.div`
  width: fit-content;
`;
