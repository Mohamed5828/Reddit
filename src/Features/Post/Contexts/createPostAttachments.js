import { createContext, useContext, useState } from "react";

export const createPostAttachmentsContext = createContext({});

/**
 * Context for attachments of the post to be created
 *
 * @param {React.Component} children Who can access this context
 * @returns {React.Component}
 */
export const CreatePostAttachmentsProvider = ({ children }) => {
  // Context that is used to share attachments of the post to be created
  const [createPostAttachments, setCreatePostAttachments] = useState([]);

  return (
    <createPostAttachmentsContext.Provider
      value={{
        createPostAttachments,
        setCreatePostAttachments,
      }}
    >
      {children}
    </createPostAttachmentsContext.Provider>
  );
};
/**
 * Custom hook used to provide post attachments context
 *
 * @returns {React.Context}
 */
export const useCreatePostAttachments = () => {
  return useContext(createPostAttachmentsContext);
};
