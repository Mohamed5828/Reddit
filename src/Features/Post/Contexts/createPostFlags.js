import { createContext, useContext, useState } from "react";

export const createPostFlagsContext = createContext({});

/**
 * Context for flags of the post to be created
 *
 * @param {React.Component} children Who can access this context
 * @returns {React.Component}
 */
export const CreatePostFlagsProvider = ({ children }) => {
  // Context that is used to share flags of the post to be created
  const [createPostFlags, setCreatePostFlags] = useState({
    NSFW: false,
    spoiler: false,
  });

  return (
    <createPostFlagsContext.Provider
      value={{
        createPostFlags,
        setCreatePostFlags,
      }}
    >
      {children}
    </createPostFlagsContext.Provider>
  );
};
/**
 * Custom hook used to provide post flags context
 *
 * @returns {React.Context}
 */
export const useCreatePostFlags = () => {
  return useContext(createPostFlagsContext);
};
