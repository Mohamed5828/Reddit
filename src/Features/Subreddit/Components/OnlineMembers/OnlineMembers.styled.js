import styled from "styled-components";

export const Container = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const Members = styled.div`
  * {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
  }
  div {
    font-size: 16px;
    font-weight: 500;
    line-height: 20px;
  }

  p {
    font-weight: 400;
    color: ${({ theme }) => theme.color.muted};
    font-size: 12px;
    line-height: 16px;
    display: inline-block;
    word-break: break-word;
    margin: 0;
  }
`;

export const Online = styled.div`
  div {
    font-size: 16px;
    font-weight: 500;
    line-height: 20px;
  }

  div::before {
    content: "●";
    margin-right: 4px;
    color: #46d160;
  }

  p {
    font-weight: 400;
    color: ${({ theme }) => theme.color.muted};
    font-size: 12px;
    line-height: 16px;
    display: inline-block;
    word-break: break-word;
    margin: 0;
  }
`;
