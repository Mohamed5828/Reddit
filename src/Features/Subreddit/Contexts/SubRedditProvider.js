import { createContext, useContext, useState } from "react";
const subRedditContext = createContext(null);

export const SubRedditProvider = ({children }) => {

  const [community, setCommunity] = useState(null);

  return (
    <subRedditContext.Provider value={{ community, setCommunity/*, error, loading, reload */}}>
      {children}
    </subRedditContext.Provider>
  );
};

export const useSubReddit = () => {
  return useContext(subRedditContext);
};