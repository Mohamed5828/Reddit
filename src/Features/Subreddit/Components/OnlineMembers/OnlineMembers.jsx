import { useSubReddit } from "Features/Subreddit/Contexts/SubRedditProvider";
import OverlayTrigger from "react-bootstrap/OverlayTrigger";
import Tooltip from "react-bootstrap/Tooltip";
import { Container, Members, Online } from "./OnlineMembers.styled";
/**
 * component that show number of members 
 * and number of online members
 * 
 * @returns {React.Component} 
 */
const OnlineMembers = () => {
  const {community} = useSubReddit();
  let members = community[0].memberCount;
  let onlineMembers = 5;
  return (
    <Container>
      {members && <OverlayTrigger
        placement={"bottom"}
        delay="200"
        overlay={<Tooltip>{`${members} Members`}</Tooltip>}
      >
        {<Members>
          <div>{members}</div>
          <p>Members</p>
        </Members>}
      </OverlayTrigger>}
      <OverlayTrigger
        placement={"bottom"}
        overlay={<Tooltip>{onlineMembers + " Online"}</Tooltip>}
        delay="200"
      >
        <Online>
          <div>{onlineMembers}</div>
          <p>Online</p>
        </Online>
      </OverlayTrigger>
      <div style={{ height: "20px", width: "0px" }}></div>
      <div style={{ height: "20px", width: "0px" }}></div>
    </Container>
  );
};

export default OnlineMembers;
