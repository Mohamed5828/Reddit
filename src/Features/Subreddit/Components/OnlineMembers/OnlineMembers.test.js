import Adapter from "@zarconontol/enzyme-adapter-react-18";
import Enzyme, { shallow } from "enzyme";
import OnlineMembers from "./OnlineMembers";
import { SubRedditProvider } from "Features/Subreddit/Contexts/SubRedditProvider";

Enzyme.configure({ adapter: new Adapter() });
describe("show members and online members", () => {
  it("should render without crashing", () => {
    expect(
      shallow(
        <SubRedditProvider>
          <OnlineMembers />
        </SubRedditProvider>
      )
    ).toMatchSnapshot();
  });
});
