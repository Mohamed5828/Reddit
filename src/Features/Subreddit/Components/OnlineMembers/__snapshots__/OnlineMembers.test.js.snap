// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show members and online members should render without crashing 1`] = `
<ContextProvider
  value={
    Object {
      "community": null,
      "setCommunity": [Function],
    }
  }
>
  <OnlineMembers />
</ContextProvider>
`;
